OpenCL_C.txt

```diff
@@ -2482,94 +2737,102 @@ p = cp; // illegal
-types shall have the same representation and alignment requirements.
-All pointers to structure types in the same address space shall have the
-same representation and alignment requirements as each other.
-All pointers to union types in the same address space shall have the same
-representation and alignment requirements as each other.
+cp = lp; // compile-time error
+cp = pp; // compile-time error
+cp = gp; // compile-time error
+----------
 
-*Clause 6.3.2.3 - Pointers, replace the first two paragraphs with the
-following paragraphs*:
+Example below demonstrates explicit conversions for pointers pointing to
+different address spaces.
 
-If a pointer into one address space is converted to a pointer into another
-address space, then unless the original pointer is a null pointer or the
-location referred to by the original pointer is within the second address
-space, the behavior is undefined.
-(For the original pointer to refer to a location within the second address
-space, the two address spaces must overlap).
+[source,c]
+----------
+#ifdef __opencl_c_generic_address_space
 
-A pointer to `void` in any address space may be converted to or from a
-pointer to any incomplete or object type.
-A pointer to any incomplete or object type in some address space may be
-converted to a pointer to `void` in an enclosing address space and back
-again; the result shall compare equal to the original pointer.
+global int *gp;
+local int *lp;
+private int *pp;
+constant int *cp;
 
-For any qualifier _q_, a pointer to a non-_q_-qualified type may be
-converted to a pointer to the _q_-qualified version of the type (but with
-the same address-space qualifier or the generic address space); the values
-stored in the original and converted pointers shall compare equal.
+int *p;
+gp = (global int *)lp; // illegal to cast between named address spaces
+p = (int *)lp; // legal to cast from global to generic
+gp = (global int*)p; // legal to cast from generic to global
 
-*Clause 6.3.2.3 - Pointers, replace the last sentence of paragraph 4 with*:
+#endif
+----------
 
-Conversion of a null pointer to another pointer type yields a null pointer
-of that type.
-Any two null pointers whose referenced address spaces overlap shall compare
-equal.
+In nested pointers implicit conversions between address spaces are disallowed.
+Explicitly casting between different address spaces in nested pointers is
+allowed but the use of such pointers can lead to incorrect behavior i.e.
+accessing invalid memory sections.
 
-*Clause 6.5.2.2 - Function calls, change the second bullet of paragraph 6
-to*:
+[source,c]
+----------
+#ifdef __opencl_c_generic_address_space
 
-both types are pointers to qualified or unqualified versions of a character
-type or `void` in the same address space or one type is a pointer in a named
-address space and the other is a pointer in the generic address space.
+kernel void mykernel(...)
+{
+local int *local * ll;
+global int *local * gl;
+int *local * nl;
```

Comment:
I think it'd be worth a few comments here to describe what's going on since this is a pretty tricky example.

Reply:
Ok, I added/changed the comment, but if you think it's not very clear I cam try to see if the example can be modified somehow...